File: src/api/analysis.ts

import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";

// Define types for analyses
export interface Analysis {
  id: number;
  project_id: number;
  agent_id: number;
  data: Record<string, any>;
  status: "pending" | "in_progress" | "completed" | "failed";
  created_at: string;
  updated_at: string;
  completed_at?: string;
}

export interface AnalysisCreate {
  project_id: number;
  agent_id: number;
  data: Record<string, any>;
}

export interface Memory {
  id: string;
  text: string;
  memory_type: string;
  tag?: string;
  score?: number;
  timestamp?: number;
  metadata?: Record<string, any>;
}

export interface QuoteSource {
  document_id?: string;
  chunk_id?: number;
  start_char?: number;
  end_char?: number;
  metadata?: {
    filename?: string;
    paragraph?: number;
    [key: string]: any;
  };
}

export interface Quote {
  text: string;
  source?: QuoteSource;
}

export interface Theme {
  name: string;
  description: string;
  keywords?: string[];
  quotes?: (string | Quote)[];
}

export interface AnalysisResults {
  themes?: Theme[];
  insights?: Array<string | {
    theme: string;
    quote?: string;
    summary?: string;
    source?: QuoteSource;
  }>;
  summary?: string;
  sentiment?: {
    overall: string;
    score: number;
    breakdown?: Record<string, any>;
  };
  memory_used?: Memory[];
  [key: string]: any;
}

// API functions
async function fetchAnalyses(projectId: number) {
  const response = await fetch(`/api/v1/analysis/project/${projectId}`);
  if (!response.ok) {
    throw new Error(`Failed to fetch analyses for project ${projectId}`);
  }
  return response.json();
}

async function fetchAnalysis(id: number) {
  const response = await fetch(`/api/v1/analysis/${id}`);
  if (!response.ok) {
    throw new Error(`Failed to fetch analysis ${id}`);
  }
  return response.json();
}

async function fetchAnalysisResults(id: number) {
  const response = await fetch(`/api/v1/analysis/${id}/results`);
  if (!response.ok) {
    throw new Error(`Failed to fetch results for analysis ${id}`);
  }
  return response.json();
}

async function createAnalysis(data: AnalysisCreate) {
  const response = await fetch("/api/v1/analysis", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });
  
  if (!response.ok) {
    throw new Error("Failed to create analysis");
  }
  
  return response.json();
}

// React Query hooks
export function useProjectAnalyses(projectId: number) {
  return useQuery({
    queryKey: ["analyses", projectId],
    queryFn: () => fetchAnalyses(projectId),
    enabled: !!projectId,
  });
}

export function useAnalysis(id: number) {
  return useQuery({
    queryKey: ["analysis", id],
    queryFn: () => fetchAnalysis(id),
    enabled: !!id,
  });
}

export function useAnalysisResults(id: number, status: string) {
  return useQuery({
    queryKey: ["analysis-results", id],
    queryFn: () => fetchAnalysisResults(id),
    enabled: !!id && status === "completed",
  });
}

export function useCreateAnalysis() {
  const queryClient = useQueryClient();
  
  return useMutation({
    mutationFn: createAnalysis,
    onSuccess: (data) => {
      queryClient.invalidateQueries({ queryKey: ["analyses", data.project_id] });
    },
  });
}

File: src/schemas/analysis.py

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime

class AnalysisBase(BaseModel):
    project_id: int = Field(..., example=1)
    agent_id: int = Field(..., example=1)
    data: Dict[str, Any] = Field(
        ..., 
        example={
            "interviews": [
                {
                    "id": "interview-1",
                    "text": "In this interview, the participant mentioned...",
                    "metadata": {"participant_id": "P001", "date": "2023-04-15"}
                }
            ],
            "parameters": {
                "theme_count": 5,
                "include_quotes": True
            }
        }
    )

class AnalysisCreate(AnalysisBase):
    pass

class AnalysisUpdate(BaseModel):
    status: Optional[str] = Field(None, example="completed")
    results: Optional[Dict[str, Any]] = None

class AnalysisResponse(AnalysisBase):
    id: int
    status: str = Field(..., example="pending")
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class MemoryItem(BaseModel):
    id: str = Field(..., example="mem-123")
    text: str = Field(..., example="Users expressed frustration with the signup process.")
    memory_type: str = Field(..., example="long_term")
    tag: Optional[str] = Field(None, example="user_feedback")
    score: Optional[float] = Field(None, example=0.92)
    timestamp: Optional[float] = Field(None, example=1679504400)
    metadata: Optional[Dict[str, Any]] = None

class QuoteSource(BaseModel):
    document_id: Optional[str] = Field(None, example="interview-1")
    chunk_id: Optional[int] = Field(None, example=3)
    start_char: Optional[int] = Field(None, example=145)
    end_char: Optional[int] = Field(None, example=267)
    metadata: Optional[Dict[str, Any]] = Field(None, example={"filename": "interview1.txt", "paragraph": 4})

class Quote(BaseModel):
    text: str = Field(..., example="I found the sign-up process frustrating and unclear.")
    source: Optional[QuoteSource] = None

class Theme(BaseModel):
    name: str = Field(..., example="User Frustration")
    description: str = Field(..., example="Users expressing negative emotions about product experience")
    keywords: Optional[List[str]] = Field(None, example=["frustrated", "confused", "difficult"])
    quotes: Optional[List[Quote]] = None

class AnalysisResults(BaseModel):
    summary: Optional[str] = None
    themes: Optional[List[Theme]] = None
    insights: Optional[List[Dict[str, Any]]] = None
    sentiment: Optional[Dict[str, Any]] = None
    memory_used: Optional[List[MemoryItem]] = None
